aea/aea.py

```diff
@@ -223,7 +229,41 @@ def _handle(self, envelope: Envelope) -> None:
             return
 
         for handler in handlers:
-            handler.handle(msg)
+            self._handle_message_with_handler(msg, handler)
+
+    def _handle_message_with_handler(self, message: Message, handler: Handler) -> None:
+        """
+        Handle one message with one predefined handler.
+
+        :param message: message to be handled.
+        :param handler: handler suitable for this message protocol.
+        """
+        with ExecTimeoutThreadGuard(self._execution_timeout) as timeout_result:
+            handler.handle(message)
+
+        if timeout_result.is_cancelled_by_timeout():
+            logger.warning(
+                "Handler `{}` was terminated as its execution exceeded the timeout of {} seconds. Please refactor your code!".format(
+                    handler, self._execution_timeout
+                )
+            )
+
+    def _behaviour_act(self, behaviour: Behaviour) -> None:
+        """
+        Call behaviour's act.
+
+        :param behaviour: behaviour already defined
+        :return: None
+        """
+        with ExecTimeoutThreadGuard(self._execution_timeout) as timeout_result:
+            behaviour.act_wrapper()
```

Comment:
what happens if someone in the `act` method does:
```python
def act(self):
    try:
        ...
    except TimeoutException:
        pass
    # do whatever I want

Reply:
@marcofavorito 
that exception will be handled. A possible solution is to make TimeoutException as a sublass of BaseException, so `except Exception:` will not catch it, but anyway user is able to catch it explicitly and make some shutdown code for this.

Reply:
yes, worth considering inheriting from `BaseException`, since it is a bad practice to do bare `except:` or `except BaseException`, and in that case the fault of bad behaviour would be of the developer.

Reply:
updated, thanks for the notice!
switched to BaseException